File: interpreters/elixir/test/support/corpus.ex
defmodule APH.TestCorpus do
  @moduledoc """
  Reads published example envelopes off disk for the suite.

  The corpus is the one this repository publishes and nothing here mints or
  signs anything new: a binding's tests exist to prove the boundary carries the
  published bytes faithfully, and a fixture invented for a test proves only
  that the test agrees with itself. Where a case needs an envelope the corpus
  does not contain — a forged label, a smuggled field — the test DERIVES it
  from a published one in view of the reader, so what changed is visible in the
  test rather than buried in a committed file.

  Two ways of naming the corpus live here on purpose. `read!/1` and the two
  named goldens address files INDIVIDUALLY, which is what a deep test needs.
  `manifest/0` and `example_files/0` address the corpus as a SET, which is what
  catches the thing individual names never can: this suite once read exactly two
  files by name and enumerated nothing, so a vector could land in `examples/`
  and the binding would report success having never opened it.

  Compiled only under `MIX_ENV=test`; see `elixirc_paths/1` in mix.exs.
  """

  # Resolved from `__DIR__` at compile time rather than from the current
  # working directory: `mix test` can be invoked from anywhere, and a suite
  # that silently read a DIFFERENT corpus than the published one would be
  # testing nothing while staying green. Four levels up is the repository root.
  @examples_dir Path.expand("../../../../examples", __DIR__)

  # The corpus INVENTORY, which is not itself a vector. Every binding's
  # enumerator skips this name for the same reason: it is the list, not a thing
  # on the list, and a suite that strict-parsed it as an envelope would fail on
  # the inventory rather than on a document anyone publishes.
  @manifest_file "manifest.json"

  @doc """
  The examples directory this suite reads, resolved once at compile time.
  """
  @spec examples_dir() :: String.t()
  def examples_dir, do: @examples_dir

  @doc """
  Reads a published example envelope by file name, raising if it is missing.

  Raising is the point: a renamed or removed example must break the suite
  loudly at the read, not quietly turn a later assertion vacuous.
  """
  @spec read!(String.t()) :: String.t()
  def read!(name) do
    @examples_dir |> Path.join(name) |> File.read!()
  end

  @doc """
  The decoded corpus manifest — the one inventory every binding measures
  against.

  Reading a directory tells you what IS there; a kept list tells you what
  SHOULD be. Only the disagreement between the two is evidence, which is why
  this module exposes both and the suite compares them. Raises if the manifest
  is missing or malformed: an inventory that cannot be read is
  indistinguishable from an empty one, and an empty one makes every comparison
  against it pass by comparing nothing to nothing.
  """
  @spec manifest() :: map()
  def manifest do
    @manifest_file |> read!() |> Jason.decode!()
  end

  @doc """
  The file names the manifest claims as the conformance corpus, sorted.
  """
  @spec conformance_files() :: [String.t()]
  def conformance_files do
    manifest() |> Map.fetch!("conformance") |> Enum.sort()
  end

  @doc """
  The manifest's excluded entries: present in the repository, deliberately
  outside the conformance claim, each carrying the one-line reason it is out.
  """
  @spec excluded_entries() :: [map()]
  def excluded_entries do
    Map.fetch!(manifest(), "excluded")
  end

  @doc """
  Every top-level `*.json` in the corpus, ENUMERATED FROM DISK rather than
  remembered, sorted, with the inventory itself skipped by name.

  Subdirectories are deliberately not descended into: the top level IS the
  conformance corpus, and what lives below it is inventoried as excluded rather
  than enumerated here.
  """
  @spec example_files() :: [String.t()]
  def example_files do
    @examples_dir
    |> File.ls!()
    |> Enum.filter(&String.ends_with?(&1, ".json"))
    |> Enum.reject(&(&1 == @manifest_file))
    |> Enum.reject(&File.dir?(Path.join(@examples_dir, &1)))
    |> Enum.sort()
  end

  @doc """
  The signed `PrincipalSigned` golden — the CHAIN arm of the proof union, and
  the form a forged label imitates.
  """
  @spec principal_signed_golden() :: String.t()
  def principal_signed_golden, do: read!("principal_signed_envelope.json")

  @doc """
  A pre-chain envelope — the SINGLE-OBJECT arm of the proof union, carrying no
  `attestationMode` at all, which the protocol reads as `NotaryAttested`.
  """
  @spec legacy_slack_reply() :: String.t()
  def legacy_slack_reply, do: read!("slack_reply_envelope.json")
end

File: interpreters/typescript/testkit/corpus.ts
/**
 * The published `examples/` corpus, and the verdict this implementation is
 * expected to return for each file.
 *
 * The table below is the comparison harness's contract. Two properties are
 * asserted by `test/corpus.test.ts` and both matter:
 *
 *  1. every row's file, when present on disk, produces the stated verdict;
 *  2. every `*.json` on disk has a row.
 *
 * Property 2 is why this file states no TOTAL. A count in prose goes stale the
 * moment a vector lands and nothing catches it; a bidirectional coverage check
 * fails loudly instead, naming the file nobody classified. Vectors this
 * repository publishes but has not committed yet are therefore listed as
 * `optional` rows rather than left out — a row that describes a file not yet
 * minted is a plan; a file with no row is a hole.
 */

import { existsSync, readFileSync, readdirSync } from 'node:fs';

import type { AphErrorCode } from '../src/errors.js';

/**
 * Finds the repository root by looking for two files that only exist there.
 *
 * Walking up beats a relative path built from `import.meta.url`: the compiled
 * tests sit at a different depth than the sources, and a hard-coded `../../..`
 * silently resolves somewhere wrong the first time the layout moves.
 */
function findRepoRoot(): URL {
  let candidate = new URL('../', import.meta.url);
  for (let depth = 0; depth < 8; depth += 1) {
    if (
      existsSync(new URL('examples/README.md', candidate)) &&
      existsSync(new URL('spec/aph-0.1.md', candidate))
    ) {
      return candidate;
    }
    candidate = new URL('../', candidate);
  }
  throw new Error(
    'could not locate the APH repository root above ' +
      `${import.meta.url} (looked for examples/README.md beside spec/aph-0.1.md)`,
  );
}

export const REPO_ROOT: URL = findRepoRoot();
export const EXAMPLES_DIR: URL = new URL('examples/', REPO_ROOT);

export function exampleUrl(fileName: string): URL {
  return new URL(fileName, EXAMPLES_DIR);
}

export function exampleExists(fileName: string): boolean {
  return existsSync(exampleUrl(fileName));
}

/** The file's TEXT, because §7.1.7.1's byte bound applies to bytes, not to a parsed value. */
export function readExample(fileName: string): string {
  return readFileSync(exampleUrl(fileName), 'utf8');
}

/** Every `*.json` in `examples/`, enumerated from disk rather than remembered. */
/**
 * The corpus INVENTORY, which lives in the corpus directory and is not itself
 * a vector. Every enumerator skips it by name: it is the one file in
 * `examples/*.json` that would fail envelope parsing for the honest reason
 * that it was never an envelope.
 */
export const MANIFEST_FILE = 'manifest.json';

export function listExampleFiles(): string[] {
  return readdirSync(EXAMPLES_DIR)
    .filter((name) => name.endsWith('.json') && name !== MANIFEST_FILE)
    .sort();
}

/**
 * What this implementation must decide about one corpus file.
 *
 *  - `admit` — full §8.3 verification succeeds.
 *  - `refuse` — full §8.3 verification fails with exactly `code`.
 *  - `offline-proofs-only` — the file's notary key does not travel inside the
 *    document and this implementation never fetches, so no key can be handed
 *    in and the assertion is narrowed to what is checkable without one:
 *    strict parse, proof structure, and every proof whose `verificationMethod`
 *    is a `did:key` and therefore carries its own key. Narrower than `admit`,
 *    and honest about which half is unchecked.
 */
export type CorpusVerdict =
  | { readonly kind: 'admit' }
  | { readonly kind: 'refuse'; readonly code: AphErrorCode }
  | { readonly kind: 'offline-proofs-only' };

export interface CorpusRow {
  readonly file: string;
  /**
   * False for a vector this repository publishes that is not committed yet;
   * the harness skips the row until the file appears on disk.
   */
  readonly required: boolean;
  readonly verdict: CorpusVerdict;
  /** WHY this file is in the corpus — what a verifier learns from getting it right. */
  readonly proves: string;
}

/**
 * SEVEN of the eight shape-only fixtures share a verdict and a reason, so they
 * share this builder: seven copies of the same sentence would be seven places
 * to update when the reason changes. The eighth,
 * `slack_new_with_extensions_envelope.json`, is spelled out as its own row
 * below because it reaches the same verdict while proving something extra.
 *
 * `APH_E001` rather than a parse error is the whole point of these rows. Their
 * `proofValue` strings are illustrative multibase, and their notary is a
 * `did:key` whose bytes this verifier decodes offline — so the document parses,
 * the structure checks out, and the SIGNATURE is what fails. A verifier that
 * reported these as malformed would be failing at step 1 and calling it step 5.
 */
function shapeOnlyFixture(file: string, channel: string): CorpusRow {
  return {
    file,
    required: true,
    verdict: { kind: 'refuse', code: 'APH_E001' },
    proves:
      `strict parse of the ${channel} recipient-addressing shape (§7.4, opaque sub-fields), ` +
      'the single-object proof carriage of §7.1.11, and that an illustrative proofValue is ' +
      'refused as a SIGNATURE failure and not as a malformed document',
  };
}

export const CORPUS: readonly CorpusRow[] = [
  shapeOnlyFixture('slack_reply_envelope.json', 'Slack thread-reply'),
  shapeOnlyFixture('email_reply_envelope.json', 'email reply'),
  shapeOnlyFixture('discord_dm_envelope.json', 'Discord direct-message'),
  shapeOnlyFixture('teams_channel_envelope.json', 'Teams channel-post'),
  shapeOnlyFixture('whatsapp_envelope.json', 'WhatsApp'),
  shapeOnlyFixture('google_chat_envelope.json', 'Google Chat space'),
  shapeOnlyFixture('imessage_envelope.json', 'iMessage'),
  {
    file: 'act_classification_envelope.json',
    required: true,
    verdict: { kind: 'refuse', code: 'APH_E001' },
    proves:
      '§7.1.12 `actClassification` parses in the INDEPENDENT implementation — the ' +
      'family-qualified labels and the pinned vocabulary reference — and that carrying ' +
      'the claim does not change the verdict, which stays the shape-only signature ' +
      'refusal. The digest is the shipped guardrail bundle\'s own, so this fixture also ' +
      'proves the two artifacts agree on what that bundle is called',
  },
  {
    file: 'audience_bound_envelope.json',
    required: true,
    verdict: { kind: 'refuse', code: 'APH_E001' },
    proves:
      '§7.1.13 `audience` parses in the INDEPENDENT implementation — the named ' +
      'recipient and the open-membered `channelBinding` with its closed `kind` — and ' +
      'that carrying it does not change the verdict, which stays the shape-only ' +
      'signature refusal reached BEFORE step 5a. The step 5a gate itself is pinned on ' +
      'minted envelopes in the roundtrip suite, where a real signature lets the ' +
      'pipeline reach it',
  },
  {
    file: 'slack_new_with_extensions_envelope.json',
    required: true,
    verdict: { kind: 'refuse', code: 'APH_E001' },
    proves:
      'all three §7.5 registered extensions parse — `appleAurAcceptance`, ' +
      '`linkedMandate.ap2SignedPayloadB64`, and `linkedMandate.vaultMutation` with its ' +
      'internally-tagged `kind` and snake_case interior (§7.5.3) — and that the extensions ' +
      'do not change the verdict, which stays the shape-only signature refusal',
  },
  {
    file: 'principal_signed_envelope.json',
    required: true,
    verdict: { kind: 'admit' },
    proves:
      'THE (→) CROSS-VERIFICATION: four real Ed25519 signatures made by the reference ' +
      'implementation verify here — both envelope proofs over their §7.2.1 bases (the ' +
      'principal base being the ONE-ELEMENT ARRAY form) and both §6.1 mandate signatures — ' +
      'plus the §7.1.11 chain linkage, the §7.2.1 issuance order, and the §7.1.7.1 ' +
      'mandate-to-envelope bindings',
  },
  {
    file: 'ts_minted_envelope.json',
    // Minted by `scripts/mint_ts_envelope.ts`; materialized at landing time.
    required: false,
    verdict: { kind: 'admit' },
    proves:
      'THE (←) CROSS-VERIFICATION ARTIFACT, read back by its own producer: this file is ' +
      'minted by this implementation and verified by the Rust reference in ' +
      "`ts_minted_cross_verify.rs`. Its row here closes the loop on the producer's side, so " +
      'a mint that drifts from this verifier fails before it ever reaches cargo',
  },
  {
    file: 'es256_signed_envelope.json',
    required: false,
    verdict: { kind: 'offline-proofs-only' },
    proves:
      'the `ecdsa-jcs-2019` MUST-support path of §8.1 against a vector this implementation ' +
      'did not mint. Narrowed to offline proofs because the vector post-dates this table and ' +
      "its notary key is not knowable here; the principal proof's did:key carries its own",
  },
  {
    file: 'detached_jws_envelope.json',
    required: false,
    verdict: { kind: 'offline-proofs-only' },
    proves:
      'the `JsonWebSignature2020` carriage of §8.2 — the fixed protected header, `b64:false` ' +
      'with `crit:["b64"]`, and the empty payload segment — against a vector this ' +
      'implementation did not mint. Narrowed for the same reason as the ES256 row',
  },
];
